Make Mac, MSW and Unix joysticks behave the same.

mickp opened the pull request:
Addresses trak#18233.

* Moved wxCtz to math.h.
* All implementations use int for m_buttonChange on button events.
* m_buttonChange has one bit set to indicate which button changed.
* Added wxJoystickEvent::GetButtonOrdinal() to return 0-based index of
button that changed.

Perhaps the wxCtz implementation should be moved elsewhere, as it is
currently in the header which may affect compile times.

mickp committed:
Make Mac, MSW and Unix joysticks behave the same.

Addresses trak#18233.

* Moved wxCtz to math.h.
* All impleme:wqntations use int for m_buttonChange on button events.
* m_buttonChange has one bit set to indicate which button changed.
* Added wxJoystickEvent::GetButtonOrdinal() to return 0-based index of
button that changed.

Perhaps the wxCtz implementation should be moved elsewhere, as it is
currently in the header which may affect compile times.

vadz commented on include/wx/math.h:
```diff
@@ -181,4 +181,24 @@ inline double wxRadToDeg(double rad) { return (rad * 180.0) / M_PI; }
 
+
+
+/* Count trailing zeros. Use optimised builtin where available. */
```
Duplicate line?

vadz commented on include/wx/math.h:
```diff
@@ -181,4 +181,24 @@ inline double wxRadToDeg(double rad) { return (rad * 180.0) / M_PI; }
+
+/* Count trailing zeros. Use optimised builtin where available. */
+/* Count trailing zeros. Use optimised builtin where available. */
+inline unsigned int wxCtz(unsigned x)
```
This should be inside `#ifdef __cplusplus` section.

vadz replied:
I'd still put it there because it's not (and shouldn't be) `extern "C"`, so it's not callable from C anyhow.

vadz commented on interface/wx/event.h:
```diff
@@ -1741,10 +1741,17 @@ class wxJoystickEvent : public wxEvent
 
+    /**
+        Returns the 0-indexed ordinal of the button changing state.
+    */
```
Please add `@since 3.1.2` here.

vadz commented on include/wx/event.h:
```diff
@@ -2760,6 +2761,7 @@ class WXDLLIMPEXP_CORE wxJoystickEvent : public wxEvent
     int GetZPosition() const { return m_zPosition; }
     int GetButtonState() const { return m_buttonState; }
     int GetButtonChange() const { return m_buttonChange; }
+    int GetButtonOrdinal() const {return wxCtz(m_buttonChange); }
```
Very minor, but please add the missing space after `{`

vadz commented on src/osx/core/hidjoystick.cpp:
```diff
@@ -872,7 +872,7 @@ void* wxJoystickThread::Entry()
             }
 
-            wxevent.SetButtonChange(nIndex+1);
+            wxevent.SetButtonChange(1 << nIndex);
```
It might be purely theoretic, but shouldn't we check for overflow here?

mickp replied:
I'll add 
```wxCHECK_RET(j_evt.number < CHAR_BIT * (int) sizeof(int), (void *) -1);```

vadz reviewed (changes requested):
Looks good, but can't be merged before `wxCtz()` is moved into the right place to fix the CI builds.

Could you please do it? TIA!

vadz commented:
The procedure you describe is correct, not sure what errors are you getting with it.

But I'm not sure we want to add another file just for this... We already define `wxGCD()` in `src/common/utilscmn.cpp`, so we could just define `wxCtz()` there too.

mickp commented:
Ah - nice one! I was looking for something like a utils source file, but must have skipped over that one.

mickp committed:
Fixes for continuous integration.

* Moved wxCtz implementation from math.h utilscmn.cpp, and renamed to
wxCTZ to match style of wxGCD.
* Added bounds checks on bit shifts in joystick implementations.
* Added version note to change in defined behaviour of GetButtonChange.
* Minor style fixes.

mickp commented:
OK - updated with fixes for continuous integration.
* Moved wxCtz implementation from math.h utilscmn.cpp, and renamed to wxCTZ to match style of wxGCD.
* Added bounds checks on bit shifts in joystick implementations.
* Added version note to change in defined behaviour of GetButtonChange.
* Minor style fixes.

mickp committed:
Declared wxCTZ for C++ only.

mickp committed:
Fix j_evt.number (linux) to nIndex (mac).

mickp commented:
For some reason, Github is no longer updating this pull request. Closing here, and resubmitted as #991.

mickp commented:
For some reason, Github is no longer updating this pull request. Closing here, and resubmitted as #991.

mickp commented:
For some reason, Github is no longer updating this pull request. Closing here, and resubmitted as #991.
